### PR TITLE
Ad lifecycle reporting

### PR DESCRIPTION
**Pull request by matamegger**
### Problem

Ad tracking is done using custom events, instead of the conviva provided ad analytics.

### Solution

- Migrate the ad tracking to the ad analytics. This PR provides a basic integration of the ad lifecycle.
- Adjust existing tests
- Add new tests for cleanup and the new ad tracking trigger 

### Notes

Finer grained tracking is implemented in follow up PRs. (To keep PR sizes smaller, or at least the production code changes)

### Checklist
- [x] I added an update to `CHANGELOG.md` file
- [x] I ran all the tests in the project and they succeed

### CODE REVIEW

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift [hunk 1]
```diff
@@ -510,39 +534,48 @@ extension ConvivaAnalytics: BitmovinPlayerListenerDelegate {
...
+        adAnalytics.reportAdStarted(adInfo)
+        adAnalytics.reportAdMetric(
+            CIS_SSDK_PLAYBACK_METRIC_PLAYER_STATE,
+            value: PlayerState.CONVIVA_PLAYING.rawValue
         )
     }
 
     func onAdFinished() {
-        videoAnalytics.reportAdBreakEnded()
+        adAnalytics.reportAdEnded()
```
Comment (matamegger) [inline BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:553]: This change is a behaviour change of the integration!!! (and of course the related changes i.e. moving the orignal call into the ad break finished event, and the starting calls in the ad started and ad break started)

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift [hunk 2]
```diff
@@ -383,6 +387,26 @@ public final class ConvivaAnalytics: NSObject {
     }
 }
 
+private extension Ad {
+    var adInfo: [String: Any] {
```
Comment (matamegger) [inline BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:391]: Will be extended in a follow up PR.

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift [hunk 3]
```diff
@@ -510,39 +534,48 @@ extension ConvivaAnalytics: BitmovinPlayerListenerDelegate {
...
         )
-        var adAttributes = [String: Any]()
-        adAttributes["c3.ad.position"] = adPosition.rawValue
-        videoAnalytics.reportAdBreakStarted(
-            AdPlayer.ADPLAYER_CONTENT,
-            adType: AdTechnology.CLIENT_SIDE,
-            adBreakInfo: adAttributes
+        adInfo["c3.ad.position"] = adPosition.rawValue
+
+        adAnalytics.reportAdLoaded(adInfo)
```
Comment (stonko1994) [inline BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:544]: On iOS we do have the `AdManifestLoaded` event 🤔 Could this be considered for `reportAdLoaded`?

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift [hunk 4]
```diff
@@ -510,39 +534,48 @@ extension ConvivaAnalytics: BitmovinPlayerListenerDelegate {
...
         )
     }
 
     func onAdFinished() {
-        videoAnalytics.reportAdBreakEnded()
+        adAnalytics.reportAdEnded()
     }
 
     func onAdSkipped(_ event: AdSkippedEvent) {
-        customEvent(event: event)
```
Comment (stonko1994) [inline BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:531]: The web integration kept the `customEvents` reporting for now to not break existing integrations.

https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/104#discussion_r1625646064

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift [hunk 5]
```diff
@@ -510,39 +534,48 @@ extension ConvivaAnalytics: BitmovinPlayerListenerDelegate {
...
+            adBreakInfo: adAttributes
+        )
     }
 
     func onAdBreakFinished(_ event: AdBreakFinishedEvent) {
-        customEvent(event: event)
+        videoAnalytics.reportAdBreakEnded()
     }
 
     func onDestroy() {
```
Comment (stonko1994) [inline BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:581]: I can't comment on `internalEndSession()`:

Should we also end the `adAnalytics` in there? See [this](https://github.com/bitmovin/bitmovin-player-web-analytics-conviva/pull/104#discussion_r1625596100) change on the web integration.

File: BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift [hunk 6]
```diff
@@ -13,6 +13,18 @@ import Foundation
 enum AdEventUtil {
     static let positionRegexPattern = "pre|post|[0-9]+%|([0-9]+:)?([0-9]+:)?[0-9]+(\\.[0-9]+)?"
 
+    static func parseAdPosition(event: AdBreakStartedEvent, contentDuration: TimeInterval) -> ConvivaSDK.AdPosition {
```
Comment (stonko1994) [inline BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift:16]: Why do we need both now? 🤔 

I don't understand why we need the information multiple times. If there is a pre-roll AdBreak, every Ad in there will be reported as pre-roll as well right?

File: Example/Tests/Doubles/BitmovinPlayerTestDouble.swift [hunk 7]
```diff
@@ -251,6 +275,47 @@ class BitmovinPlayerTestDouble: BitmovinPlayerStub, TestDoubleDataSource {
     }
 }
 
+class TestAdBreak: NSObject, AdBreak {
+    var identifier: String
+    var scheduleTime: TimeInterval
+    var ads: [any Ad]
```
Comment (stonko1994) [inline Example/Tests/Doubles/BitmovinPlayerTestDouble.swift:281]: This `any` should not be needed. Ad is not a generic type. **applies for all**

```suggestion
    var ads: [Ad]
```

**Review by stonko1994 — COMMENTED** [review]
Can you merge `develop` into this branch. I tried something on the CI to stabilize the unit test execution (not the flaky tests)

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift (hunk 3 above)
Comment (matamegger) [reply BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:544]: The same exists on Android, but this is about the media file/stream itself.

File: BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift (hunk 6 above)
Comment (matamegger) [reply BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift:16]: The documentation is not very clear on this part. The Android integration only track metadata in the `AdStarted`.

Revisiting this, I will drop this again.

File: Example/Tests/Doubles/BitmovinPlayerTestDouble.swift (hunk 7 above)
Comment (matamegger) [reply Example/Tests/Doubles/BitmovinPlayerTestDouble.swift]: https://github.com/bitmovin/bitmovin-player-ios-analytics-conviva/pull/68/commits/9a465cd6a083bf712e1661d08117504f48d5933b

File: BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift (hunk 6 above)
Comment (matamegger) [reply BitmovinConvivaAnalytics/Classes/Helper/AdEventUtil.swift]: https://github.com/bitmovin/bitmovin-player-ios-analytics-conviva/pull/68/commits/c85be19d7758a462fcbcb383a04c49dc75e39884
https://github.com/bitmovin/bitmovin-player-ios-analytics-conviva/pull/68/commits/49b1f8a998af3c6c74b357e53d469bbbd2f73dd6

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift (hunk 4 above)
Comment (matamegger) [reply BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift:531]: https://github.com/bitmovin/bitmovin-player-ios-analytics-conviva/pull/68/commits/1b42d71be7815a588413001c82e1ea94d74a5464
https://github.com/bitmovin/bitmovin-player-ios-analytics-conviva/pull/68/commits/3de0a9aef58d8a4d022048ef779c6ab0ff481032

File: BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift (hunk 5 above)
Comment (matamegger) [reply BitmovinConvivaAnalytics/Classes/ConvivaAnalytics.swift]: I am not sure. I couldn't find anything in that regard in the documentation.
I also looked at another integration, and this is not done there.

**Pull request merged**